### PR TITLE
Add duration coverage found missing during review of PR #3232

### PR DESCRIPTION
**Pull request by wolenetz**
Original author of some of this new coverage is @tidoust.
This commit replaces those in PR #3232, which had become highly
conflicting with fixes to mediasource-duration.html and others which
landed while #3232 was in review.

This commit also includes further coverage beyond #3232.

### CODE REVIEW

**Comment by wolenetz** [comment]
@tidoust @jyavenard Please take a look ASAP. Note that Chromium will not yet pass some of these since https://crbug.com/639144 is not yet fixed. Specifically, each of the following currently triggers the old async remove behavior in Chrome (and hence, the subtests fail):

* Test setting same duration multiple times does not fire duplicate durationchange
* Test setting the duration to less than the highest starting presentation timestamp will throw
* Truncating the duration throws an InvalidStateError exception when new duration is less than the highest buffered range start time of one of the track buffers
* Truncating the duration throws an InvalidStateError exception when new duration is less than a buffered coded frame presentation time

**Comment by wolenetz** [comment]
This PR is intended to replace #3232

**Comment by jyavenard** [comment]
The value don't appear good to me, in particular for the video track.
The video you're adding is 2s long only or close to, so setting the new duration to a value close to 2 will be a problem.

It happens to be okay with mp4, because the last frame that still fit with the append window happened to also be the last sample of the file and have an end time of 2.066666 so you'll have your assert.


With webm however:
FAIL: Truncating the duration throws an InvalidStateError exception when new duration is less than a buffered coded frame presentation time
assert_throws: function "function () { mediaSource.duration = newDuration; }" did not throw
after the audio appendBuffer we have the sourceBufferAudio.buffered equal to (0.808000, 1.783000)

The reason for 1.783 end time is that we have an overlapping frame:
with a start time of 1.784000 and end time of 1.807000
Per spec:
9. If frame end timestamp is greater than appendWindowEnd, then set the need random access point flag to true, drop the coded frame, and jump to the top of the loop to start processing the next coded frame.
so that frame is dropped.

For video however we have the last frame of the file with a start time of 1.967000 and end time of 2.000333

setting duration to 2 will succeed because it's greater than the highest *starting* time

So I suggest the following:
do not set sourceBufferVideo.appendWindowEnd.
set new duration to a value greater than the sourceBufferAudio.appendWindowEnd : say 1.9

Or:
have a specific check for webm and mp4.
webm: must fail if new duration is < 1.967000
mp4: must fail if new duration is < 2.033333

the later is probably more thorough.

Be careful that due to chrome issue of misusing dts, the last mp4 frame has a dts of 1.966666, that complicates the matter if you attempt to verify the issue with chrome or edge.

Tested with Firefox Nightly, which is fully up to date with the current specs.
You can test yourself with going into about:support and setting the preference:
media.mediasource.mp4.enabled to false
This will enable the webm mediasource by default.

I should point out that running current master with just this patch I see failures there for:
Test seek starts on duration truncation below currentTime
Test appendBuffer completes previous seek to truncated duration
Test endOfStream completes previous seek to truncated duration

each due to "assert_equals: Playback time is truncatedDuration while seeking expected 1.522 but got 1.527"
The test incorrectly assume that seeking will seek to the truncatedDuration, however as duration is set to the highest end time across all tracks, that it seeks to 1.527 is correct.

The tests haven't been updated for the last duration change spec change from a few days ago.

**Comment by wolenetz** [comment]
I'll take a closer look in the morning here; I think
s/appendWindowStart/timestampOffset/ is what I was intending to do but
forgot in the new subtest.

**Comment by wolenetz** [comment]
Also, good point about the need to update the current time while seeking to
truncated duration expectation. I'll fix that too in an update to this pr.

**Comment by wolenetz** [comment]
@plehegar FYI I'm working on updating this at the moment.

**Comment by wolenetz** [comment]
In c21af2b, using timestampOffset instead of appendWindowStart appears to work as expected.
In Chrome, even with the current ISO-BMFF DTS/PTS conflation there:
Buffered ranges prior to attempting duration truncation and verifying the exception:
ISO-BMFF:
audio:0.8-1.8
video:1.5-2.866666

no ISO-BMFF support: (custom local build of Chrome):
audio:0.8-1.797
video:1.5-3

I'll now work on fixing the currentTime expectation while seeking to truncated duration.

File: media-source/mediasource-duration.html [hunk 1]
```diff
@@ -53,7 +58,9 @@
 
                   test.waitForExpectedEvents(function()
                   {
-                      // remove will not remove partial frames. The truncated duration is as such the highest end time.
+                      // remove will not remove partial frames. The truncated duration is as such the highest end time
```
Comment (wolenetz) [inline media-source/mediasource-duration.html]: this is messy. I've cleaned it up in 1724173

**Comment by wolenetz** [comment]
OK. As of 1724173, I believe this PR is ready to merge. @plehegar I'll merge this shortly so that implementation reports may be regenerated ASAP.
@jyavenard @jdsmith3000 We can clean this up in a later PR if I've regressed something.

**Comment by jyavenard** [comment]
LGTM for both webm and mp4.

However, I don't think we're actually testing the particular of setting the duration to a time within the duration of the last sample.
Pedantic I know, but seeing that there's a special case in the spec for it, verifying the behaviour would be good

**Pull request merged**